pkg/scheduler/scheduler_test.go

```diff
@@ -0,0 +1,273 @@
+)
+
+const (
+	crpName = "test-crp"
+
+	bindingName    = "test-binding"
+	altBindingName = "another-test-binding"
+
+	policySnapshotName        = "test-policy-snapshot"
+	altPolicySnapshotName     = "another-test-policy-snapshot"
+	anotherPolicySnapshotName = "yet-another-test-policy-snapshot"
+)
+
+var (
+	ignoreObjectMetaResourceVersionField = cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ResourceVersion")
+	ignoreTypeMetaAPIVersionKindFields   = cmpopts.IgnoreFields(metav1.TypeMeta{}, "APIVersion", "Kind")
+)
+
+// TestMain sets up the test environment.
+func TestMain(m *testing.M) {
+	// Add custom APIs to the runtime scheme.
+	if err := fleetv1beta1.AddToScheme(scheme.Scheme); err != nil {
+		log.Fatalf("failed to add custom APIs to the runtime scheme: %v", err)
+	}
+
+	os.Exit(m.Run())
+}
+
+// TestCleanUpAllBindingsFor tests the cleanUpAllBindingsFor method.
+func TestCleanUpAllBindingsFor(t *testing.T) {
+	now := metav1.Now()
+	crp := &fleetv1beta1.ClusterResourcePlacement{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              crpName,
+			DeletionTimestamp: &now,
+			Finalizers:        []string{fleetv1beta1.SchedulerCRPCleanupFinalizer},
+		},
+	}
+
+	bindings := []*fleetv1beta1.ClusterResourceBinding{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: bindingName,
+				Labels: map[string]string{
+					fleetv1beta1.CRPTrackingLabel: crpName,
+				},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: altBindingName,
+				Labels: map[string]string{
+					fleetv1beta1.CRPTrackingLabel: crpName,
+				},
+			},
+		},
+	}
+
+	fakeClient := fake.NewClientBuilder().
+		WithScheme(scheme.Scheme).
+		WithObjects(crp, bindings[0], bindings[1]).
+		Build()
+	// Construct scheduler manually instead of using NewScheduler() to avoid mocking the controller
+	// manager.
+	s := &Scheduler{
+		client:         fakeClient,
+		uncachedReader: fakeClient,
+	}
+
+	ctx := context.Background()
+	if err := s.cleanUpAllBindingsFor(ctx, crp); err != nil {
+		t.Errorf("cleanUpAllBindingsFor() = %v, want no error", err)
+	}
+
+	if err := fakeClient.Get(ctx, types.NamespacedName{Name: crpName}, crp); err == nil {
+		t.Errorf("Get() CRP = %v, want no error", err)
+	}
+	wantCRP := &fleetv1beta1.ClusterResourcePlacement{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              crpName,
+			DeletionTimestamp: &now,
+			Finalizers:        []string{},
+		},
+	}
+	if diff := cmp.Diff(crp, wantCRP, ignoreObjectMetaResourceVersionField, ignoreTypeMetaAPIVersionKindFields); diff != "" {
+		t.Errorf("updated CRP diff (-got, +want): %s", diff)
+	}
+
+	if err := fakeClient.Get(ctx, types.NamespacedName{Name: bindingName}, bindings[0]); err == nil || !errors.IsNotFound(err) {
+		t.Errorf("Get() binding %s = %v, want not found", bindingName, err)
+	}
+	if err := fakeClient.Get(ctx, types.NamespacedName{Name: altBindingName}, bindings[1]); err == nil || !errors.IsNotFound(err) {
+		t.Errorf("Get() binding %s = %v, want not found", altBindingName, err)
+	}
```

Comment:
use list func to check the number of bindings instead?

Reply:
This is a little bit overkill IMO but I guess the more tests the merrier.

Reply:
Done.